### ISSUE TITLE
Add unicode shortcut ctrl+Shift+u support

### ISSUE THREAD

**Issue by Jeshwanth** [jwilm/alacritty#866]
Which operating system does the issue occur on?

Linux

If on linux, are you using X11 or Wayland?
X11

In Terminal App, I can enter the unicode value by typing ctrl+Shift+u but in alacritty its not available, can you please add this ?

**Comment by justfortherec** [comment]
This works for me with version 0.1.0 on Ubuntu 16.04 LTS.
(this is in a bash shell running in tmux)

When pressing Ctrl+Shift+u, a small gray pop-up shows what Unicode char I am typing.

**Comment by Chais** [comment]
I have the same problem. ibus-daemon is running, `CTRL`+`SHIFT`+`U` works for the browser, but not alacritty. But since it doesn't work for thunar either it might well be a misconfiguration on my part.

**Comment by chrisduerr** [comment]
I don't think this is due to misconfiguration, but rather that @justfortherec has popup which is unusual. Alacritty renders only single characters without making use of any library for layout so that's probably why this doesn't work.

**Comment by towc** [comment]
This feature is almost a must for me. I don't have much experience in rust, but how can I help land it?

**Comment by chrisduerr** [comment]
Right now I'm not even sure what the best approach for landing this would be. The only idea I have would be making use of the text rendering used for the FPS counter to display the unicode information at the bottom of the terminal.

Then we could have a terminal shortcut for unicode input (let's say ctrl+shift+u by default) which opens this menu at the bottom (like a `Unicode: ` that shows up) and then the user can type characters which are then displayed in there. Once `enter` is pressed the symbols typed could be converted to a single unicode symbol.

I'm not quite sure if this is the best, or even a reasonable way to do this, but it seems like the easiest way to implement this feature without having to interfer with the actual terminal output. What do you think about this @jwilm?

**Comment by chrisduerr** [comment]
Either way will not have a performance impact on the normal use because it's gated behind the keyboard shortcut which has to be pressed until any of the logic kicks in.

I don't think it's a good idea to not display it at all, making it work with the interface used for the FPS counter should be fairly simple, so that shouldn't be a problem. Maybe it would even work to get it directly inside the terminal but I'm not too sure about that.

**Comment by francesca64** [comment]
This doesn't look like it's part of the XIM protocol? https://en.wikipedia.org/wiki/Unicode_input#In_X11_(Linux_and_Unix_variants)

The best thing to do is to read the GTK source code, as much fun as that sounds. GLFW is lighter reading material, but this sounds like something they probably didn't implement (GLFW isn't as complete as you might think; the owner of kitty will tell you as much).

I can't say I'll be able to look into this soon. Those macOS keyboard improvements were just something I did on a whim after having to test another PR on macOS, since I was too lazy to reconnect my keyboard and mouse to my Linux machine (this is the true story of why I completely rewrote keyboard handling). I'm actually supposed to be working on proper HiDPI support, and after that there's the major event loop API change that will require structural changes to every backend, so that will keep me busy.

**Comment by chrisduerr** [comment]
Thanks for exploring this! I think that there would be some simple ways to directly handle this in alacritty, so that might be our best bet!

**Comment by dvergeylen** [comment]
Hi,

I am afraid it doesn't work anymore: running Linux here (X11), commit 8d5e122.
Typing <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>u</kbd> doesn't prompt unicode pop up anymore.

Can anyone confirm this?

**Comment by chrisduerr** [comment]
> I am afraid it doesn't work anymore

This issue is still open, so I'm not aware of this getting fixed. Was this feature working for you previously?

**Comment by justfortherec** [comment]
> I am afraid it doesn't work anymore: running Linux here (X11), commit [8d5e122](https://github.com/jwilm/alacritty/commit/8d5e1226d500579692128439abd9c991ba9a72d3).
Typing Ctrl + Shift + u doesn't prompt unicode pop up anymore.

The pop-up doesn't show anymore for me either. However, the unicode character is still inserted correctly if I blindly type it; only the pop-up is not visible.

> Was this feature working for you previously?

It used to work for me but does not anymore. I don't know if it worked any time since I upgraded from Ubuntu 16.04 (Unity desktop) to 18.04 (Gnome desktop). I know for sure it worked before the upgrade and I thought it also worked after (I am not 100% positive though).

**Comment by chrisduerr** [comment]
It would be much appreciated if you could test an old version to check if this was broken by Alacritty. Though it's totally possible that this was due to a change in Ubuntu (not necessarily the fault of Ubuntu, but causing a separate bug to be triggered).

**Comment by dvergeylen** [comment]
Hi reverted to 879ea05, which I am sure used to work for me (had a IME displaying unicode dynamically). It doesn't work anymore so I guess problem is on OS side.

Just like @justfortherec, I moved from Ubuntu 18.04 to Debian testing (X11) recently and now don't see an IME anymore (sadly).

Don't hesitate to ask for tests @chrisduerr, would be pleased to help. 👍

**Comment by chrisduerr** [comment]
Thanks for testing! I'd assume that Ubuntu might have switched around their IME system to something winit does not yet properly handle 100%.

**Comment by dvergeylen** [comment]
https://github.com/tomaka/winit/pull/772 might be related :thinking:.
They plan to ship its merge in winit 0.19.1 once released (see https://github.com/tomaka/winit/pull/798).

**Comment by dvergeylen** [comment]
This worked again after installing `ibus` on Debian bullseye:

```shell
sudo apt install ibus
```
![Capture d’écran du 2019-11-12 17-36-32](https://user-images.githubusercontent.com/8804879/68690907-1df20680-0573-11ea-9a91-7025fc7c433d.png)
![Capture d’écran du 2019-11-12 17-36-52](https://user-images.githubusercontent.com/8804879/68690913-20546080-0573-11ea-9ab0-b3775c993691.png)

IME is not aligned with prompt line, but that's another story 😉

**Comment by pwnorbitals** [comment]
Does it work well for everyone now ? If so, as it has been a year since last activity, we could close this issue

**Comment by tinywrkb** [comment]
@pwnorbitals what's working?  
`Ctrl+Shift+u`? this doesn't seem to be supported.

IME frameworks? I'm on an updated build from git master, Fcitx5, Sway 1.5, and this still doesn't work in Wayland even though #2734 is supposedly fixed. No issue in X11 but I never had a problem with X11.

**Comment by kchibisov** [comment]
fcitx5 on sway won't work without sway patch to grab input, from alacritty side everything for some basic IME stuff is done, and it works on GNOME Wayland.

**Comment by tinywrkb** [comment]
@kchibisov thanks for your reply. This is indeed working with [Sway's input grab patch](https://github.com/swaywm/sway/pull/4932) pulled in, with a git master build of course.  
Fcitx5 the word selection window doesn't appear but it might be a Fcitx5 issue.

**Comment by kchibisov** [comment]
>Fcitx5 the word selection window doesn't appear but it might be a Fcitx5 issue.

Yeah, and we don't support inline input method right now, which is #1613

**Comment by jas01** [comment]
Hi @kchibisov,  so as I understand the classic «ctrl+shift+u» not going to work soon. 
Just like to know what would be the best way to input some UTF-8 by the code like «shift+ctrl+u+1f603» give 😃 with alacritty
Thanks

**Comment by iGEL** [comment]
@chrisduerr wrote on on Apr 22, 2018:

> Then we could have a terminal shortcut for unicode input (let's say ctrl+shift+u by default) which opens this menu at the bottom (like a `Unicode: ` that shows up) and then the user can type characters which are then displayed in there. Once `enter` is pressed the symbols typed could be converted to a single unicode symbol.

I would really prefer if it could be done like GTK+ does it (QT does it the same way, I think). So, hold `ctrl` and `shift` and then tapping `u` plus the unicode hex code, release `shift` and `ctrl`. 

What's displayed is less important to me personally, as I want to program my keyboard send this key sequence with one button. GTK+ actually just displays the typed characters with an underscore, so I see `UDC` until I release shift & ctrl, then that gets replaced with `Ü`. Perfect is the enemy of the good, so I think, that or even don't display anything at all before releasing shift & ctrl is better than it's now: not working. ;-)

**Comment by akiross** [comment]
I'm also interested in this feature and I also like the inline display à la GTK (`u` followed by digits, all underlined, which gets confirmed with enter or space).

I know it's not much, but as a workaround for people needing **a few** fixed unicode glyphs, I'm using the compose key and customizing the `.XCompose` file.

**Comment by koopa1338** [comment]
Did someone already look into this? I would also appreciate what @iGEL mentioned to just press `ctrl` + `shift` + `u` and then send my Unicode. I would like to help if possible but don't know where to start with this, may be someone already know how GTK is doing this and can share some insights.

**Comment by chrisduerr** [comment]
> Did someone already look into this?

No. There's questionable utility behind this considering the existence of IME.

**Comment by alerque** [comment]
The Compose key isn't working on Wayland either. So basically ya'll just want us to type ASCII only?

**Comment by kchibisov** [comment]
I'm not sure what is not working for you, I use compose key every day on Wayland. To type whatever I want, like 🤔 emojis. There're also IMEs for Wayland doing exactly ctrl+shift+u.

You setup Compose key in your compositor and then it should just work.

**Comment by alerque** [comment]
I have the compose key setup in my compositor config and it works almost everywhere *except* Alacritty (and anther winit based apps like Neovide). It works just fine here in Firenvim inside Firefox. It sounds like you might have your Compose key bound to an IME function rather than acting as a compose key.

**Comment by kchibisov** [comment]
No, it's a raw compose key which is handled by libxkbcommon, you could have locale issue since compose key is bound to it and its behavior, the locale is used for example to find out the compose tables. If your everywhere is GTK than it's not representative since they bundle some compose tables.

**Comment by alerque** [comment]
Do you have any resources for what I should be looking for? I know what $LANG and /etc/locale.conf are setup to and `localectl status` seems fine.

**Comment by kchibisov** [comment]
Run `alacritty -vv` from the terminal, iirc it will throw the warnings if you have issues with locale and compose on Wayland. I don't remember how it looks exactly though. Maybe you don't have compose tables installed, since some distros split the package. If you need more real time help just you could ask me on irc instead of pinging issue.

The `LC_ALL=en_US.UTF-8 alacritty` could also be a good test.

**Comment by alerque** [comment]
Wow now we're on to something. Setting `LC_ALL` does in fact bring Compose capabilites to Alacritty! Here's the thing though: I'm pretty sure that's *incorrect* behavior on the part of Alactritty. LC_ALL should **not** be used for this in normal usage because it overrides all other variables. It's literally the one thing `localectl` will refuse to set in `/etc/locale.conf` (https://wiki.archlinux.org/title/Locale#LC_ALL:_troubleshooting). My system has a whole plethera of `LC_*` variables available:

```console
$ locale
LANG=en_US.utf-8
LC_CTYPE="en_US.utf-8"
LC_NUMERIC="en_US.utf-8"
LC_TIME="en_US.utf-8"
LC_COLLATE="en_US.utf-8"
LC_MONETARY="en_US.utf-8"
LC_MESSAGES="en_US.utf-8"
LC_PAPER="en_US.utf-8"
LC_NAME="en_US.utf-8"
LC_ADDRESS="en_US.utf-8"
LC_TELEPHONE="en_US.utf-8"
LC_MEASUREMENT="en_US.utf-8"
LC_IDENTIFICATION="en_US.utf-8"
LC_ALL=
```

Looking at the output of `alacritty -vv` it looks like xkbcommon is where this is being thrown. Does that mean I should go chase this down in a library or is Alacritty responsible for passing this on?

**Comment by kchibisov** [comment]
`LC_ALL` is for debugging purposes, we don't use it, maybe you don't have `utf-8` files lowercased or LANG is wrong, maybe you've forgot to generate locale of some sort? Given that you've tested `UTF-8` and it worked, I'd suggest to change LANG to `en_US.UTF-8` instead of `en_US.utf-8`.

Anyway, your issue is with the way you've set your locale.

**Comment by nhatzHK** [comment]
Every other program I use (that is aimed at linux users) supports entering unicode with ctrl+shift+u. Is there anything blocking implementing this? The issue is opened since 2017. Do you need help to get this done? Or is there another way to get the same behavior in Alacritty? I would have just switched to a different terminal, but this is the first and only issue I've encountered with Alacritty so far.

**Comment by towc** [comment]
> Is there anything blocking implementing this? 

it's already implemented and has been working for a while now, out of the box. Pressing ctrl+shift+u brings up this interface for me, where I can type the code and end it as usual.

![image](https://github.com/alacritty/alacritty/assets/6382186/79c36217-9707-4691-ab8e-7b20d2c1b546)

It looks like some users are still encountering configuration issues, so I recommend reading through the previous comments and applying others' fixes and seeing if that does it.

I'm not sure why this issue is still open.

**Comment by chrisduerr** [comment]
> Is there anything blocking implementing this?

Yeah it's just completely unnecessary. There are standard composable mechanisms for inputting arbitrary unicode in interactive fashion (IME). The ctrl+shift+u mechanism is just a poorly designed semi-standard that people have gotten used to.

**Comment by ibotty** [comment]
I agree that ctrl+shift+u is a semi-standard, but it's the only you have to send a unicode character from a programmable keyboard.

**Comment by kchibisov** [comment]
>I agree that ctrl+shift+u is a semi-standard, but it's the only you have to send a unicode character from a programmable keyboard.

Have you heard about XCompose, IME, etc? There're a lot of ways to do that besides the `ctrl+shuft+u`. ibus I think binds by default to do such a thing. I know that on macOS you need special API to do that, but it could be supported as a binding in the future since you need to bring that menu from the app.

**Comment by ibotty** [comment]
Maybe we have a misunderstanding.  I have a, say hotkey, on my keyboard that wants to type e.g. 👍 directly, without any menu opening and ideally without much configuration on the host OS.  

Ctrl+shift+u does support this to some degree.  AFAICT Compose does not.  I'll have to look (again) into IME whether it supports non-interactive input.

**Comment by kchibisov** [comment]
If you're using libxkbcommon (linux) you can define whatever you want in `.XCompose`, like:

```
include "%L"
<Multi_key> <h> <m> : "🤔"
<Multi_key> <plus> <1> : "👍"
<Multi_key> <minus> <1> : "👎"
<Multi_key> <h> <o> <o> : "🎉"
<Multi_key> <c> <r> <a> <b> : "🦀"
<Multi_key> <s> <r> <g> : "¯\\_(ツ)_/¯"
```

And then just send a sequence which may do compose.

¯\_(ツ)_/¯ 🦀 🤔

**Comment by kchibisov** [comment]
And no, ctrl+shift+u is not os agnostic, on apple you're facing popup menu with emojis, etc, and ctrl+shift+u requested here is a picker and how we do picker if we do it is alacritty specific, keep in mind that there're IMEs working exactly like ctrl+shift+u.

**Comment by ibotty** [comment]
I don't care about Windows or Mac (If I had to work with them I would use something different in the keyboard firmware), but multiple Linux computers.  Using XCompose is a setup I'll have to replicate everywhere.  The point is, that ctrl+shift+u works now in multiple Desktop environments (GNOME and wlroots based ones), but not in alacritty.

**Comment by kchibisov** [comment]
> The point is, that ctrl+shift+u works now in multiple

it's a non-standard application thing, it doesn't work for me in all the apps I have, usually on gnome it should work oob if you use ibus and you have it set to handle ctrl+shift+u. It worked on X11 last time I've checked.

**Comment by ibotty** [comment]
I don't dispute that it's semi-standard or that it's poorly designed.  Really, no objection at all.  It does work on the firefox flatpak out of the box though (even using sway and Sericea without any configuration re ibus).

AFAICT there is no better alternative for that (narrow!) use case though, is there?

**Comment by nhatzHK** [comment]
FWIW use IBus https://wiki.archlinux.org/title/IBus#Unicode_input

**Comment by anarcat** [comment]
i want to voice my interest in adding a ctrl-shift-u shortcut to alacritty. it's not a feature i use very often, but it's *very* useful when i need it. i also think it's standard  enough that alacritty should follow it, or at least provide a way for users to configure it.

the foot(1) terminal emulator made the switch [three years ago](https://codeberg.org/dnkl/foot/issues/1183) or so and while there were some complaints about the keybinding change (control-shift-u was for opening URLs too), they smoothed out, and i think everyone is on the same page there now (and people who don't can just change the keybinding).

I'll repeat what i stated there, three years ago:

> oh! unicode! my favorite topic to not have to fix except i always have to because i'm not a native english speaker yay! :p 
> 
> i just stumbled upon this as well. most apps out there do use control-shift-u and i'm trying to standardize my desktop to do so as well. i *thought* this was a standard, but it turns out, of course not:
> 
> https://en.wikipedia.org/wiki/Unicode_input#Hexadecimal_input
> 
> TL;DR: on a Mac, it's <kbd>Option</kbd> (<kbd>⌥</kbd>) and the numbers. On Windows, it's <kbd>alt</kbd> then <kbd>+</kbd> on the numeric keypad and then the digits. Or just <kbd>alt</kbd> and the digits. Or you need to edit the registry. Who knows. The only "standard" in there is GTK, Qt, and ChromeOS all agreeing on <kbd>control shift u</kbd>.
> 
> (Amazingly, each *app* has its own way too: Emacs is <kbd>control x 8</kbd>, vim is <kbd>Control V u</kbd>, Libreoffice uses <kbd>Alt</kbd>, Opera is <kbd>Control-Shift-x</kbd>... diversity. yay.)
> 
> So I would argue for using at least one of the existing conventions. I've personnally rebound <kbd>control-shift-u</kbd> and assigned browse-urls elsewhere:
> 
> ```
> [key-bindings]
> show-urls-launch=Control+Shift+o
> unicode-input=Control+Shift+u
> ```
> 
> <kbd>Control+Shift+o</kbd> makes sense to me for that keybinding because <kbd>o</kbd> is a mnemonic for "open url" and wasn't used by anything else I could think of.

So, to put this more strongly than I did above: in Linux, there *is* an (adhoc) standard: GTK, Qt, and ChromeOS all use control-shift-u. In this text area where i'm typing now (Firefox), control-shift-u works out of the box, and i suspect a *lot* of people expect this to work.

I think it would just be polite for new users to do the same. And while this might not seem like an important thing for english speakers who only need ASCII, it's one of those things us non-native speakers end up using quite often, especially when we type on someone else's keyboard for some reason. :)

If we had, at least, some function to input the code and I could rebind the keybindings, it's be pretty happy...

Thanks for considering it!

**Comment by olafur-palsson** [comment]
Hi, I just found this issue from a different issue I have. I use Ctrl+Shift+^ to go to a previous file in neovim and it doesn't work.

From testing it looks like Ctrl+Shift shortcuts don't work in general for me (Wayland Plasma 6.5.2 based on Ubuntu 24.04). These shortcuts work on Konsole as well as Ctrl+Shift+& outputting the same character as Shift+& (% in my case).

This make me think that this issue could be related to how Ctrl+Shift are handled in general.
